### ISSUE TITLE
Failed to write file '~/.nextflow/plugins/nf-amazon-1.0.1' to plugins folder

### ISSUE THREAD

**Issue by drpatelh** [nextflow-io/nextflow#1859]
## Bug report 

A number of users of the nf-core/rnaseq pipeline have now reported this bug. It appears that the `nf-amazon-1.0.1` plugin is downloaded to a temp directory but Nextflow fails to copy it from there to `~/.nextflow/plugins/`

Please see https://github.com/nf-core/rnaseq/issues/552#issuecomment-763232497

This only seems to be a problem on certain infrastructures, I haven't experienced it personally.

**Comment by abhi18av** [comment]
This issue has been observed here as well 
https://github.com/ctmrbio/BACTpipe/pull/141#issuecomment-762781346

I'm reproducing the error trace mentioned in that comment 

```
N E X T F L O W ~ version 21.01.1-edge
Launching /proj/uppstore2017270/Emilio/bactpipe-testing/tests/210119/BACTpipe/bactpipe.nf [pedantic_raman] - revision: a8340c3f3b
Downloading plugin nf-amazon@1.0.1
Failed to write file '/home/emilir/.nextflow/plugins/nf-amazon-1.0.1' to plugins folder

And from .nextflow.log:

Jan-19 08:54:10.467 [main] ERROR nextflow.cli.Launcher - @unknown
org.pf4j.PluginRuntimeException: Failed to write file '/home/emilir/.nextflow/plugins/nf-amazon-1.0.1' to plugins folder
at nextflow.plugin.PluginUpdater.download0(PluginUpdater.groovy:170)
at nextflow.plugin.PluginUpdater.access$0(PluginUpdater.groovy)
at nextflow.plugin.PluginUpdater$_safeDownload_closure2.doCall(PluginUpdater.groovy:188)
at nextflow.plugin.PluginUpdater$_safeDownload_closure2.call(PluginUpdater.groovy)
at nextflow.file.FileMutex.lock(FileMutex.groovy:107)
at nextflow.plugin.PluginUpdater.safeDownload(PluginUpdater.groovy:188)
at nextflow.plugin.PluginUpdater.load0(PluginUpdater.groovy:217)
at nextflow.plugin.PluginUpdater.installPlugin(PluginUpdater.groovy:140)
at nextflow.plugin.PluginUpdater.prepareAndStart(PluginUpdater.groovy:88)
at nextflow.plugin.PluginsFacade.start(PluginsFacade.groovy:194)
at nextflow.plugin.PluginsFacade.start(PluginsFacade.groovy:199)
at nextflow.plugin.PluginsFacade.setup(PluginsFacade.groovy:161)
at nextflow.plugin.Plugins.setup(Plugins.groovy:40)
at nextflow.cli.CmdRun.run(CmdRun.groovy:268)
at nextflow.cli.Launcher.run(Launcher.groovy:475)
at nextflow.cli.Launcher.main(Launcher.groovy:657)
Caused by: java.nio.file.AtomicMoveNotSupportedException: /scratch/17938909/pf4j-update-downloader5690633583307575560/nf-amazon-1.0.1 -> /home/emilir/.nextflow/plugins/nf-amazon-1.0.1: Invalid cross-device link
at sun.nio.fs.UnixCopyFile.move(UnixCopyFile.java:394)
at sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:262)
at java.nio.file.Files.move(Files.java:1395)
at nextflow.plugin.PluginUpdater.download0(PluginUpdater.groovy:167)
... 15 common frames omitted
```

**Comment by pditommaso** [comment]
Thanks for reporting. Should be this 

https://github.com/nextflow-io/nextflow/blob/bf0938268b422dcd72d14e6deff5ce035ef92a3e/modules/nf-commons/src/main/nextflow/plugin/PluginUpdater.groovy#L167

**Comment by boulund** [comment]
I just hit this issue on a new system as well (working in an NFSv4 mount). What is the easiest workaround for this issue?

Just updated to the latest edge release but no cigar there, I guess the above commit has yet to make it into a new release.

**Comment by pditommaso** [comment]
Already patched on master branch, I'm going to upload a new edge release in 48h

**Comment by boulund** [comment]
Awesome, moving fast as always!

In the meantime, if anyone else encounters this, here's how I worked around it:

1. Go to your temp directory, in my case `/tmp`. It will contain a file called `nextflow-plugin-nf-amazon-1.0.1.lock`.
2. Find the temporary folder into which the plugin was decompressed, in my case this was `pf4j-update-downloader9461589285179136600`. This contains the `nf-amazon-1.0.1` folder. 
3. Move the `nf-amazon-1.0.1` folder to your nextflow plugins folder, in my case this is `~/.nextflow/plugins`.

**Comment by pditommaso** [comment]
Should be solved with version `21.02.0-edge`.
